File: Bigger_Wordplay/solutions/get_all_compound_words.py
"""
Solution 5: What are all of the compound words? These are words made 
up of 2 smaller words. For example, “SNOWMAN” is a compound word made 
from “SNOW” and “MAN”, and “BEACHBALL” is a compound word made from 
“BEACH” and “BALL”.
"""

# Q1: Are "compound words" constructed from individual words in sowpods.txt?


# Create full set of all possible permutations of sowpods (len 2 -> AA + AAH, AAH + AA)

# Loop through dataset and check for membership against ^


# Rolling hash -> look this up

from typing import Set
from pathlib import Path


SCRABBLE_PATH: Path = Path("Bigger_Wordplay/tests/fixtures/sowpods.txt")


def get_all_compound_words(path: Path) -> Set[str]:
    matches: Set[str] = set()

    # Loop over each word in the dataset file
    with open(path, mode="r") as f:
        data: Set[str] = set(f.read().splitlines())

    # Loop over each word word in the dataset
    for word in data:
        # Pointer for sliding substring window
        pointer: int = 1
        # Iterate until pointer reaches end of given word
        while pointer < len(word):
            # Slice to discover first word if it exists
            curr_compound: str = word[0:pointer]
            if curr_compound in data:
                # Slice to discover second word
                second_compound: str = word[pointer:]
                if second_compound in data:
                    matches.add(word)
                    # Move to next word (matches tend to be compound
                    # in multiple ways)
                    break
            pointer += 1

    return matches


if __name__ == "__main__":
    matches: Set[str] = get_all_compound_words(SCRABBLE_PATH)
    for word in matches:
        print(word)

File: Bigger_Wordplay/solutions/get_all_compound_words.go
// Solution 5: What are all of the compound words? These are words made
// up of 2 smaller words. For example, “SNOWMAN” is a compound word made
// from “SNOW” and “MAN”, and “BEACHBALL” is a compound word made from
// “BEACH” and “BALL”.
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func main() {
	// Load the file into a hash map set implementation
	dataMap := loadFileIntoMapSet("Bigger_Wordplay/tests/fixtures/sowpods.txt")

	// Extract all the compound words from the data
	compoundWords := getAllCompoundWords(dataMap)

	// Print the compound words to std
	for _, word := range compoundWords {
		fmt.Println(word)
	}
}

// Load file into hash map (set)
func loadFileIntoMapSet(path string) map[string]bool {
	// Load data from a file into memory
	file, fErr := os.Open(path)
	check(fErr)
	defer file.Close()

	// Instantiate a scanner with the file
	scanner := bufio.NewScanner(file)

	// Instantiate a hash map to store our data in
	dataMap := make(map[string]bool)

	// Loop over each word and add it to the hash map
	for scanner.Scan() {
		word := scanner.Text()
		dataMap[word] = true
	}

	// Error handling for scanner
	check(scanner.Err())

	return dataMap
}

// Discover and extract all of the compound words in a dataset
func getAllCompoundWords(dataMap map[string]bool) []string {
	// Instantiate an empty map for our matches
	matches := make(map[string]bool)

	// Loop over each word in the hash map
	for word := range dataMap {
		pointer := 1
		// Scan the word to discover the first substring compound
		for pointer < len(word) {
			currCompound := word[0:pointer]
			// Check for presence in data hash map
			present, _ := dataMap[currCompound]
			if present {
				// Extract the second potential compound
				secondCompound := word[pointer:]
				// Check for presence in data hash map
				present, _ = dataMap[secondCompound]
				if present {
					matches[word] = true
				}
			}
			pointer += 1
		}
	}
	// Define a slice to hold the compound words
	compoundWords := make([]string, 0, len(matches))

	// Extract the keys and add them to the slice
	for key := range matches {
		compoundWords = append(compoundWords, key)
	}
	return compoundWords
}

// Generic error handling helper
func check(err error) {
	if err != nil {
		log.Fatal(err)
	}
}
